src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -103,21 +111,38 @@ private static BigDecimal[] makePower10Table() {
    * see {@link DataValidationUtil#isAllowedSemiStructuredType}.
    *
    * @param input Object to validate
-   * @return JSON tree representing the input
+   * @return Minified JSON string
    */
-  private static JsonNode validateAndParseSemiStructuredAsJsonTree(
+  private static String validateAndParseSemiStructured(
       String columnName, Object input, String snowflakeType, final long insertRowIndex) {
     if (input instanceof String) {
-      String stringInput = (String) input;
+      final String stringInput = (String) input;
       verifyValidUtf8(stringInput, columnName, snowflakeType, insertRowIndex);
-      try {
-        return objectMapper.readTree(stringInput);
-      } catch (JsonProcessingException e) {
+      final StringBuilderWriter resultWriter = new StringBuilderWriter(stringInput.length());
+      try (final JsonParser parser = factory.createParser(stringInput);
+          final JsonGenerator generator = factory.createGenerator(resultWriter)) {
+        while (parser.nextToken() != null) {
+          final JsonToken token = parser.currentToken();
+          if (token.isNumeric()) {
+            // If the current token is a number, we cannot just copy the current event because it
+            // would write token the token from double (or big decimal), whose scientific notation
+            // may have been altered during deserialization. We want to preserve the scientific
+            // notation from the user input, so we write the current numer as text.
+            generator.writeNumber(parser.getText());
+          } else {
+            generator.copyCurrentEvent(parser);
+          }
+        }
+      } catch (JsonParseException e) {
         throw valueFormatNotAllowedException(
             columnName, snowflakeType, "Not a valid JSON", insertRowIndex);
+      } catch (IOException e) {
+        throw new SFException(e, ErrorCode.IO_ERROR, "Cannot create JSON Parser or JSON generator");
       }
+      return resultWriter.toString();
     } else if (isAllowedSemiStructuredType(input)) {
-      return objectMapper.valueToTree(input);
+      JsonNode node = objectMapper.valueToTree(input);
```

Comment:
curious, why don't we need a similar approach in non-string input fields? please add a comment too.

The streaming reader/writer being used just for String is apt for all kinds of data, so ideally we shouldn't need this DOM-parse of the json anywhere now?

Reply:
When the users ingest a `java.lang.String` and the target column is `variant`, `array` or `object`, we have to validate it, minify it, and store it in the data file, which is what the previous block with `JsonParser` and `JsonGenerator` does. 

On the other hand, If the user passes another supported Java type, e.g. `Map<String, Object>`, `List<Object>`, `LocalDateTime` or many other types - see the list [here](https://docs.snowflake.com/user-guide/data-load-snowpipe-streaming-overview#supported-java-data-types)), we need to serialize it into a JSON string, which we then store in the data file. In this case there is nothing to parse, we have to serialize the Java object input into String, which is why we use the intermediate AST.